File: backend/app/schemas/feed.py
"""Feed API schemas (TICKET-402)."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.core.feed_constants import FEED_PAGE_SIZE_DEFAULT, FEED_PAGE_SIZE_MAX
from app.schemas.neighborhood import FeedNeighborhoodSummary


class FeedLocation(BaseModel):
    latitude: float
    longitude: float


class FeedAuthor(BaseModel):
    type: str
    id: uuid.UUID
    display_name: str
    username: str | None = None
    logo_url: str | None = None


class FeedOfferMeta(BaseModel):
    partner_offer_id: uuid.UUID
    valid_from: datetime | None = None
    valid_until: datetime | None = None
    offer_type: str | None = None
    is_flash: bool = False
    flash_ends_at: datetime | None = None
    remaining_hours: int | None = None
    remaining_minutes: int | None = None


class FeedEventMeta(BaseModel):
    local_event_id: uuid.UUID
    starts_at: datetime
    ends_at: datetime | None = None
    location_name: str
    district: str | None = None
    event_type: str | None = None
    interested_by_me: bool = False


class FeedPostItem(BaseModel):
    id: uuid.UUID
    type: str
    author: FeedAuthor
    city: str | None
    title: str | None
    body: str | None
    media_url: str | None = None
    location: FeedLocation | None = None
    like_count: int
    comment_count: int
    liked_by_me: bool
    offer: FeedOfferMeta | None = None
    event: FeedEventMeta | None = None
    neighborhood_summary: FeedNeighborhoodSummary | None = None
    created_at: datetime
    updated_at: datetime


class FeedListResponse(BaseModel):
    items: list[FeedPostItem]
    next_cursor: str | None = None


class FeedListParams(BaseModel):
    cursor: str | None = None
    limit: int = Field(default=FEED_PAGE_SIZE_DEFAULT, ge=1, le=FEED_PAGE_SIZE_MAX)

File: frontend/packages/types/src/feed.ts
/** Citizen feed types (TICKET-402 / TICKET-403) — alignés sur l’API backend. */

export type FeedPostType = "post" | "offer" | "event" | "partner_creator";

export type FeedAuthorType = "citizen" | "organization";

export type FeedReportReason = "spam" | "inappropriate" | "other";

export interface FeedLocation {
  latitude: number;
  longitude: number;
}

export interface FeedAuthor {
  type: FeedAuthorType;
  id: string;
  display_name: string;
  username: string | null;
  logo_url: string | null;
}

export interface FeedEventMeta {
  local_event_id: string;
  starts_at: string;
  ends_at: string | null;
  location_name: string;
  district: string | null;
  event_type: string | null;
  interested_by_me: boolean;
}

/** Discrete territorial badge on feed cards (TICKET-602). */
export interface FeedNeighborhoodSummary {
  slug: string;
  display_name: string;
}

export interface FeedCreatorContentMeta {
  partner_creator_content_id: string;
  published_at: string | null;
}

export interface FeedOfferMeta {
  partner_offer_id: string;
  valid_from: string | null;
  valid_until: string | null;
  offer_type: string | null;
  is_flash?: boolean;
  flash_ends_at?: string | null;
  remaining_hours?: number | null;
  remaining_minutes?: number | null;
}

export interface FeedPost {
  id: string;
  type: FeedPostType;
  author: FeedAuthor;
  city: string | null;
  title: string | null;
  body: string | null;
  media_url: string | null;
  location: FeedLocation | null;
  like_count: number;
  comment_count: number;
  liked_by_me: boolean;
  offer: FeedOfferMeta | null;
  event: FeedEventMeta | null;
  creator_content: FeedCreatorContentMeta | null;
  neighborhood_summary: FeedNeighborhoodSummary | null;
  created_at: string;
  updated_at: string;
}

export interface FeedListResponse {
  items: FeedPost[];
  next_cursor: string | null;
}

export interface FeedListParams {
  cursor?: string | null;
  limit?: number;
}

export interface PostCreatePayload {
  author_type?: "citizen";
  body: string;
  media_url?: string | null;
}

export interface FeedComment {
  id: string;
  post_id: string;
  user_id: string;
  author_display_name: string;
  author_username: string | null;
  body: string;
  created_at: string;
  updated_at: string;
}

export interface CommentListResponse {
  items: FeedComment[];
  next_cursor: string | null;
}

export interface CommentCreatePayload {
  body: string;
}

export interface ReportPostPayload {
  reason: FeedReportReason;
}
